runtime/lua/vim/net.lua

```diff
@@ -0,0 +1,364 @@
+---             - follow_redirects boolean|nil Whether to follow redirects. Defaults to true.
+---             - data string|table|nil Data to send with the request. If a table, it will be
+---             JSON-encoded. vim.net does not currently support form encoding.
+---
+---             - on_complete fun(response: table)|nil Callback function when request is
+---             completed successfully. The response has the following keys:
+---                 - ok boolean Whether the request was successful (status within 2XX range).
+---                 - headers fun(): table<string, string> Function returning a table of response headers.
+---                 - body fun(): string|nil Function returning response body. If method was HEAD,
+---                 this is nil.
+---                 - json fun(opts: table|nil): table Read the body as JSON. Optionally accepts
+---                 opts from |vim.json.decode|. Will throw errors if body is not JSON-decodable.
+---                 - method string The http method used in the most recent HTTP request.
+---                 - status number The numerical response code.
+---                 - size number The total amount of bytes that were downloaded. This
+---                 is the size of the body/data that was transferred, excluding headers.
+---                 - http_version number HTTP version used in the request.
+---             - on_err fun(err: string[])|nil An optional function recieving a `stderr_buffered` string[] of curl
+---             stderr. Without providing this function, |fetch()| will automatically raise an error
+---             to the user. See |on_stderr| and `stderr_buffered`.
+---@return number jobid A job id.
+---
+--- Example:
+--- <pre>lua
+--- -- GET a url
+--- vim.net.fetch("https://example.com/api/data", {
+---   on_complete = function (response)
+---     -- Lets read the response!
+---
+---     if response.ok then
+---       -- Read response body
+---       local body = response.body()
+---     else
+---   end
+--- })
+---
+--- -- POST to a url, sending a table as JSON and providing an authorization header
+--- vim.net.fetch("https://example.com/api/data", {
+---   method = "POST",
+---   data = {
+---     key = value
+---   },
+---   headers = {
+---     Authorization = "Bearer " .. token
+---   },
+---   on_complete = function (response)
+---     -- Lets read the response!
+---
+---     if response.ok then
+---       -- Read JSON response
+---       local table = response.json()
+---     else
+---
+---     -- What went wrong?
+---     vim.print(response.status)
+---   end
+--- })
+--- </pre>
+function M.fetch(url, opts)
+  vim.validate({
+    opts = { opts, 'table', true },
+  })
+
+  opts = opts or {}
+
+  -- Ensure that we dont download to a file in fetch()
+  opts.download_location = nil
+
+  local args = createCurlArgs(url, opts)
+
+  local out = {}
+
+  if opts._dry then
+    return args
+  end
+
+  local job = vim.fn.jobstart(args, {
```

Comment:
Is there a reason you are using vim.fn.jobstart over vim.loop.spawn?

Reply:
No